agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -71,9 +72,10 @@ def deploy_rev_changed?
     # @param event [Docker::Event]
     def on_container_event(topic, event)
       if @container && event.id == @container.id
-        debug "container event: #{event.status}"
+        at = Time.at(event.time_nano / 10**9)
+        debug "container event: #{event.status} at #{at.utc.xmlschema(9)}"
         @container_state_changed = true
-        handle_restart_on_die if event.status == 'die'.freeze
+        handle_restart_on_die if event.status == 'die'.freeze && at > @container.started_at
```

Comment:
This works to fix #2781 and ignore the `die` event from before the `restart`... not yet entirely sure if it's safe in terms of reliably picking up die events from after the restart, though... it really depends on how the Docker engine actually timestamps the container state vs the events?

Reply:
The timestamp comparisons seem to work:

```
I, [2017-11-23T11:36:14.344253 #1]  INFO -- Kontena::Workers::ServicePodWorker: restarting redis-1
I, [2017-11-23T11:36:15.089439 #1]  INFO -- Kontena::ServicePods::Restarter: restarting service: /redis-1
D, [2017-11-23T11:36:16.813080 #1] DEBUG -- Kontena::Workers::ServicePodWorker: skip redis-1 state sync at 2017-11-23 11:34:39 UTC: unchanged
D, [2017-11-23T11:36:16.813569 #1] DEBUG -- Kontena::Workers::ServicePodWorker: redis-1 stale container event: kill at 2017-11-23T11:36:15.182030439Z < 2017-11-23T11:36:16.518394890+00:00
D, [2017-11-23T11:36:16.813749 #1] DEBUG -- Kontena::Workers::ServicePodWorker: redis-1 stale container event: die at 2017-11-23T11:36:15.338782310Z < 2017-11-23T11:36:16.518394890+00:00
D, [2017-11-23T11:36:16.813986 #1] DEBUG -- Kontena::Workers::ServicePodWorker: redis-1 stale container event: stop at 2017-11-23T11:36:15.885792016Z < 2017-11-23T11:36:16.518394890+00:00
D, [2017-11-23T11:36:16.814162 #1] DEBUG -- Kontena::Workers::ServicePodWorker: redis-1 container event: start at 2017-11-23T11:36:16.551548719Z
D, [2017-11-23T11:36:16.814329 #1] DEBUG -- Kontena::Workers::ServicePodWorker: redis-1 container event: restart at 2017-11-23T11:36:16.571505308Z
```